### DISCUSSION TITLE
How Renderer works?

### DISCUSSION THREAD

**Discussion by guguxh** [wxyz-abcd/node-haxball#29] [General]
I wondering how Renderer works, and how can be implemented. Can you give me some tips?

**Comment by wxyz-abcd** [comment]
It's quite complicated for now. I almost finished web frontend project. I will put it on GitHub after some time, then you can see how it works. But the main issue is not complicated logic, it is the variable names. You may have to wait until I open-source basro's codes completely, then everybody will understand everything. :)

**Comment by wxyz-abcd** [comment]
But basically, you have to have a render function inside a Renderer object. It will get a "extrapolatedRoomPhysicsObj" object that contains all information about the game's current(extrapolated) state. You can get any information and render it however you want into wherever you want, but it would usually be a canvas, since we are using window.requestAnimationFrame. A very basic test could be like this:
```
function Renderer(canvas){
  this.c = canvas.getContext("2d", { alpha: false });
  var that = this;
  this.render = function(extrapolatedRoomPhysicsObj){ // render logic here. called inside requestAnimationFrame callback
    // you have to use the objects inside extrapolatedRoomPhysicsObj to get game information to draw here.
    // that.c.clearRect(0, 0, 1000, 1000);
    that.c.beginPath();
    that.c.rect(0, 0, 100, 100);
    that.c.fillStyle = "#ff00ff";
    that.c.fill();
  };
}

var gameCanvas = document.getElementById("canvas"); // you need a <canvas id="canvas" /> HtmlElement inside DOM for this.
var api = new API.Haxball({
  // .... ,
  renderer: new Renderer(gameCanvas)
});
// ....
```

Remember that you need to be inside a room, and the game has to be active in order to render anything.
